trade-finance-manager-api/src/v1/services/deal-cancellation/deal-cancellation.service.test.ts

```diff
@@ -0,0 +1,118 @@
+import { TfmDealCancellation } from '@ukef/dtfs2-common';
+import { CANCEL_DEAL_FUTURE_DATE, CANCEL_DEAL_PAST_DATE } from '../../../constants/email-template-ids';
+import sendTfmEmail from '../send-tfm-email';
+import { DealCancellationService } from './deal-cancellation.service';
+
+const mockPimEmailAddress = 'pim@example.com';
+const mockFormattedFacilities = 'mock formatted facilities';
+const ukefDealId = 'ukefDealId';
+const ukefFacilityIds = ['aFacilityId'];
+
+jest.mock('../send-tfm-email');
+jest.mock('../../api', () => ({
+  findOneTeam: jest.fn(() => ({ email: mockPimEmailAddress })),
+  findOneDeal: jest.fn(() => ({ dealSnapshot: { ukefDealId } })),
+  findFacilitiesByDealId: jest.fn(() =>
+    ukefFacilityIds.map((ukefFacilityId) => ({
+      facilitySnapshot: {
+        ukefFacilityId,
+      },
+    })),
+  ),
+}));
+
+jest.mock('./helpers/format-facility-ids', () => ({
+  formatFacilityIds: jest.fn(() => mockFormattedFacilities),
+}));
+
+const dealId = 'dealId';
+
+describe('sendDealCancellationEmail', () => {
```

Comment:
Can we test the other methods as well please?

Reply:
Sorry -- I knew I said I'd set up a call to chat about our testing approaches, but this week has got away with me.

Are you sure we need to be testing private functions? It feels slightly odd to expose them from the service for the sake of only testing, as they are then accessible anywhere in the service?

I generally think of private methods as 'part of the unit being tested' -- and they are only really broken from the main function for readability (as apposed to say, business logic etc)

If we were to make all our private methods public, I'd expect to mock them otherwise are we not testing the behaviour twice -- ie

```
public static anOtherwisePrivateMethod () {}

public static aPublicMethodThatUsesPrivateMethod() {}
```

Are we wanting to test `anOtherwisePrivateMethod ` standalone, as well as part of in `aPublicMethodThatUsesPrivateMethod`? Or are we mocking it in `aPublicMethodThatUsesPrivateMethod`?

I'd lean to only testing private methods once -- in the public methods they are used in as I think this is one `unit` to test?

I'll pop in a meeting for Monday now Abhi is back!

Reply:
I thought we agreed before that every function should be tested

Maybe we should convert this to pure functions so it's not a class with public/private methods

Reply:
I think my confusion is about whether we then duplicate test efforts on this in the tests for the function where it's called as well?

Reply:
Not sure why we'd want to duplicate test efforts, I would expect something like this:

```ts
describe('deal-cancellation.service', () => {
  describe('sendDealCancellationEmail', () => {});

  describe('getFacilityIds', () => {});

  describe('submitDealCancellation', () => {
    describe('it should call api.findOneDeal', () => {});

    describe('it should call api.findFacilitiesByDealId', () => {});

    describe('it should call sendDealCancellationEmail', () => {});
  });
});
```

Reply:
Ah -- In this case we align that `sendDealCancellationEmail` should be mocked as it's async... But if `sendDealCancellationEmail` would we still look to mock it in our tests as well as test it? I think this is where my main confusion around this lies... ie:

```
describe('deal-cancellation.service', () => {
  describe('sendDealCancellationEmail', () => {
    describe('it should do whatever send deal cancellation email does when successful', () => {});

    describe('it should do whatever send deal cancellation email does when successful (another thing the function does)', () => {});

    describe('it should do whatever error throwing deal cancellation email does when it fails', () => {});

    describe('it should do whatever error throwing deal cancellation email does when it fails in another way', () => {});
});

  describe('getFacilityIds', () => {});

  describe('submitDealCancellation', () => {
    describe('it should call api.findOneDeal', () => {});

    describe('it should call api.findFacilitiesByDealId', () => {});

    describe('it should do whatever send deal cancellation email does when successful', () => {});

    describe('it should do whatever send deal cancellation email does when successful (another thing the function does)', () => {});

    describe('it should do whatever error throwing deal cancellation email does when it fails', () => {});

    describe('it should do whatever error throwing deal cancellation email does when it fails in another way', () => {});
  });
});
```

Reply:
This is what i'm thinking - always open to suggestions though

```ts
describe('deal-cancellation.service', () => {
  describe('sendDealCancellationEmail', () => {
    it('should call api.findOneTeam', () => { });

    describe('when X date is Y', () => {
      it('should call sendTfmEmail with template no.1', () => { });
    });

    describe('when X date is Z', () => {
      it('should call sendTfmEmail with template no.2', () => { });
    });
  });

  describe('getFacilityIds', () => {
    it('should return an array of mapped facility IDs without null entries', () => { });
  });

  describe('submitDealCancellation', () => {
    it('should call api.findOneDeal', () => {});

    it('should call api.findFacilitiesByDealId', () => {});

    it('should call sendDealCancellationEmail', () => {});

    describe('when sendEmail errors', () => {
      it('should call throw an error', () => { });
    });
  });
});
```

Maybe the email function could/should be moved into it's own directory and then we could call/mock like `dealCancellationEmail.send();`, but don't want to introduce another new pattern.